### ISSUE TITLE
Unify `-eth1-endpoints` and `--execution-endpoints`

### ISSUE THREAD

**Issue by paulhauner** [sigp/lighthouse#3069]
## Description

[`eth1`]: https://github.com/sigp/lighthouse/tree/stable/beacon_node/eth1
[`execution_layer`]: https://github.com/sigp/lighthouse/tree/stable/beacon_node/execution_layer

We presently have two flags for specifying the connection the EE:

- `--eth1-endpoints`: present since beacon-genesis, used for downloading deposit contract data in the [`eth1`][] crate.
- `--execution-endpoints`: new in bellatrix/merge, used to specify a JWT-protected JSON RPC endpoint in the [`execution_layer`][] crate.

I think the ideal scenario is that we unify the HTTP JSON RPC code among the [`eth1`][]  and [`execution_layer`][] crates so all requests go via a JWT-protected JSON RPC endpoint. Then, we can make `--eth1-endpoints` and alias to `--execution-endpoints`. Then, when users migrate they just need to:

1. Change the port of the `--eth1-endpoints` value to point to the JWT-authenticated port (e.g., Geth's `--authrpc.port`).
1. Add the `--jwt-secrets` flag, pointing at the same file as the EE (e.g., Geth's `--authrpc.jwtsecret`).

However, we can't do this until all the EE impls have released versions with JWT support. Otherwise, all of the current "eth1" calls will start failing due to a lack of JWT. So, I think the best strategy for now is to maintain the two flags. However, I'm open to hear other opinions.

**Comment by paulhauner** [comment]
Thinking about this some more, I think I prefer a different strategy. I prefer deprecating `eth1-endpoints` in the long-term.

The path to deprecation has three phases:

1. Have `eth1-` and `execution-` endpoints as distinct flags, both required for merge networks.
   - This is where we are now.
   - Non-merge networks (mainnet) still require non-auth support, since it's not stable/prevalent on EE impls yet.
   - Merge networks *must* have auth support.
   - Our [`eth1`](https://github.com/sigp/lighthouse/tree/stable/beacon_node/eth1) crate doesn't support auth yet.
2. Allow users to supply `execution-` only (i.e., `eth1-` is option).
    - To achieve this, we need to refactor the [`eth1`](https://github.com/sigp/lighthouse/tree/stable/beacon_node/eth1) crate to optionally use the auth functionality from [`execution_layer`](https://github.com/sigp/lighthouse/tree/stable/beacon_node/execution_layer).
3. Make `eth1-` illegal and require all users to use `execution-`
    - We do this when the Bellatrix release is imminent and it doesn't make sense to run without a merge-ready EE anymore.

**Comment by mrabino1** [comment]
while not a direct issue for Lighthouse, most certainly the EE updates will require installation documentation to be updated (on both the CL an EL side, even just firewall configs may be impacted).  I understand this is a bit fluid right now; however I do stress to the team that updating docs in parallel will prove to be valuable as it is guaranteed this information doesn't disseminate Re setup at the pace we think it should.

**Comment by realbigsean** [comment]
We may still want to allow users the option to sync the deposit cache separately from their execution engine.  This could be important because the deposit cache requires log history that isn't required by the execution engine otherwise. I'm assuming Infura will still make non-engine API endpoints available, so users could still use Infura specifically for the deposit cache.

**Comment by paulhauner** [comment]
> We may still want to allow users the option to sync the deposit cache separately from their execution engine. This could be important because the deposit cache requires log history that isn't required by the execution engine otherwise. I'm assuming Infura will still make non-engine API endpoints available, so users could still use Infura specifically for the deposit cache.

I hadn't considered this, it's a good point. There is some more complexity, though. I have started to think along the lines of ensuring that the "EE" and the "eth1" should be the same nodes. It's a fairly remote scenario, but if we have a case where those nodes are different then we risk the scenario where they're on different chains. This could result in voting in an `eth1_data` that isn't compatible with the `execution_payload`s in the chain.

Along these lines, I'd been considering adding in a check to ensure that whenever we vote upon an `eth1_data` we also ensure that the `eth1_data.block_hash` is also present in one of the `execution_payload.block_hashes` in the canonical chain. It would have some edge-cases around the transition, but it would (practically) ensure that we never vote in an `eth1_data` that conflicts with a prior one. Voting in conflicting `eth1_data` is one of the scenarios that can break beacon chain liveness :scream:

**Comment by paulhauner** [comment]
I have thought through this and come up with some use-cases and changes required to achieve them. I have already done *some* of this work in the abandoned #3160, but it's not necessarily the same as this.

## Overview

### Use case 1: Pre-merge

The user wants to use multiple eth1 nodes:

`lighthouse bn --eth1-endpoints http://localhost:8545,http://shared-backup.com`

### Use case 2: Post-merge

The user wants the "eth1" functionality to happen over the same JWT-authenticated connection as the engine API and they don't want to repeat themselves via `--eth1-endpoints`:

`lighthouse bn --execution-endpoint http://localhost:8551 --execution-jwt ./jwt.key`


## Changes to CLI Flags

### Objectives

1. Update our CLI so it no longer supports multiple execution endpoints (without removing internal support from the `execution_layer` crate, we can do that later).
1. Allow users to specify `--execution-endpoint` and `--execution-jwt` flags *without* also needing `--eth1-endpoints`. This is how things should operate *after* the merge.
1. Allow users to specify `--eth1-endpoints` separately and continue to use multiple eth1 nodes (i.e., maintain pre-merge functionality).

### Changes

1. Rename the `--execution-endpoints` flag to `--execution-endpoint` (remove plural)
    - Change it so it only takes a single value.
    - Add an alias to the old `--execution-endpoints` value, so it's backwards compatible (unless multiple values are supplied)
1. Rename the `--jwts-secrets` flag to `--execution-jwt`
    - Change it so it only takes a single value.
    - Add an alias to the old `--jwts-secrets` value, so it's backwards compatible (unless multiple values are supplied)
    - The rename from "jwt-secrets" to "execution-jwt" is so that related flags are close together in alphabetical sort.
1. Rename `--jwt-id` to `--execution-jwt-id`
    - Add alias to `--jwt-id` for full backwards compat.
1. Rename `--jwt-version` to `--execution-jwt-version`
    - Add alias to `--jwt-version` for full backwards compat.
    
## Changes to internal functionality

### Objectives

1. Allow our current "eth1" crate to communicate with an eth1 node (EE) with JWT authentication (as per the engine API). This is how things should work post-merge.
1. Maintain current support for non-JWT comms to retain pre-merge functionality.

### Changes

1. Move the RPC functions in [`eth1/src/http.rs`](https://github.com/sigp/lighthouse/blob/stable/beacon_node/eth1/src/http.rs) into [`execution_layer/src/engine_api/http.rs`](https://github.com/sigp/lighthouse/blob/stable/beacon_node/execution_layer/src/engine_api/http.rs).
1. From the `eth1` crate, import the RPC functionality by importing [`execution_layer/src/engine_api/http.rs`](https://github.com/sigp/lighthouse/blob/stable/beacon_node/execution_layer/src/engine_api/http.rs).
1. Allow the `eth1` crate to take an *optional* JWT token.

**Comment by mrabino1** [comment]
Humor the basic question.. Will the execution-endpoint serve the same content as that what would have been served with port 8545 (but with additional content capabilities like block production)? thx.

**Comment by pawanjay176** [comment]
> Will the execution-endpoint serve the same content as that what would have been served with port 8545

Yes that's correct. It will be a jwt-authenticated endpoint over which the execution engine will expose the engine namespace and all other namespaces that are exposed over the non-auth http endpoint (eth, net etc)

**Comment by paulhauner** [comment]
Resolved via #3215 :tada:
